Update Rules::Taxon/Product handling of invalid match policies

jordan-brough opened the pull request:
This is extracted from discussion in https://github.com/solidusio/solidus/pull/1113.
The goal is to make invalid match_policy values less likely and to
have consistent handling of invalid values should they occur.

For `Rules::Product`: We now raise an error for invalid match policies. This was already
happening in `Rules::Product#actionable?`, it just wasn't happening in
`Rules::Product#eligible`.

For `Rules::Taxon`: We now log a warning and assume it was meant to be an `any`
match policy.  I'm being more lenient here because `Rules::Taxon` had a bug
with the default value for match policies and assuming `any` is essentially
what the previous logic did.

Also:  Add validations to help avoid invalid values in the first place.
Also:  Fix the default value for Rules::Taxon#preferred_match_policy.

See individual commits.

jordan-brough commented:
Note:  I'm going to rework this a bit.  I'm going to have the "else" clauses make an assumption instead of generating an error, while still logging the invalid value.

jrochkind commented:
how will it decide what assumption to make? Different for each rule, as it is presently?

jordan-brough commented:
I've updated the logic and updated the PR description.

@jrochkind yes, I tried to follow the previous handling for the most part.  See the updated PR description.

jhawthorn commented:
:+1:

tvdeyen commented:
👍
